### PR TITLE
refactor(rslint_parser): Specific List nodes

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

Implementing the shape validation described in #1867 requires that it's possible to tell the required shape of a node by just looking at its kind. This hasn't been possible so far for list nodes that all use the same ` LIST` kind. The kind doesn't expose any information of what the expected type of the child nodes is nor if it is a separated or a node list. 

This PR replaces the generic `LIST` kind with specific list types, for example, it introduces the `JsArrayElementList` for the list storing the elements of an array expression. 

A consequence of having explicit `LIST` nodes is that these should now also implement the `AstNode` interface, they are nodes after all. The consequences of doing so are:

* `AstSeparatedList` and `AstNodeList` must be traits that each specific list implement together with the `AstNode` interface
* It's no longer possible to use a `missing` placeholder to represent an empty list because a concrete node is required to implement `AstNode.syntax()`. 

Overall, the change seems to nicely align with how the rest of the AST facade works and makes list nodes less special, they're just regular nodes that on top implement additional methods that allow iterating over their elements. 

The main downside of this change is that it's now required to implement the `AstSeparatedList` and/or `AstNodeList` traits if someone wants to use any of the list methods. I guess, that's just how Rust works and shouldn't be too much of a surprise.

Part of #1725 #1867 

## Test Plan

Verified the changed rast files.

### CODE REVIEW

**Review by ematipico — APPROVED** [review]
I can see that it's mostly code generation and updated snapshots changes!

It seems good to me, I think we can merge it!

File: xtask/js.ungram [hunk 1]
```diff
@@ -304,17 +308,33 @@ JsAnyExpression =
 
 
 // the backtick is just a placeholder, see parser in rslint_parser/src/ast/expr_ext.rs
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply xtask/js.ungram:310]: Nah,  you're right. It's not the final grammar but certainly more than a placeholder.

File: xtask/src/codegen/ast.rs [hunk 2]
```diff
@@ -60,34 +49,51 @@ fn make_ast(grammar: &Grammar) -> AstSrc {
...
-fn handle_alternatives(grammar: &Grammar, rule: &Rule) -> Option<Vec<String>> {
+enum RuleType {
+	Union(Vec<String>),
+	Node,
+	Unknown,
+	List {
+		separated: bool,
+		element_name: String,
+	},
+}
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply xtask/src/codegen/ast.rs]: Sure. I can maybe rename it as well. It's more high level than `Rule`, it tells us whatever the NodeRule is a `Union`, `Node`, `Unknown` or `List` node.

File: xtask/js.ungram [hunk 3]
```diff
@@ -1087,7 +1135,9 @@ TsNamespaceExportDecl = 'export' 'as' 'namespace' ident: Ident? ';'?
 Specifier = name:JsName 'as'? manual__alias: JsName?
 
 // @ematipico this one is not entirely correct I think..
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply xtask/js.ungram:1130]: Yeah, kind of. I fixed some issues with exports but there could be more. Let's leave it until we re-did the export parsing.

**Pull request merged**